### PR TITLE
Tester improvements

### PR DESCRIPTION
**Pull request by Wescoeur**

### CODE REVIEW

**Review by benjamreis — APPROVED** [review]

File: setup.cfg [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [pycodestyle]
 max-line-length=120
-ignore=E261,E302,E305
+ignore=E261,E302,E305,W503
```
Comment (stormi) [inline setup.cfg:3]: Should probably belong in a separate commit with an explanation of what this warning is

File: tests/storage/lvm/test_lvm_sr.py [hunk 2]
```diff
@@ -1,5 +1,4 @@
 import pytest
-import subprocess
```
Comment (stormi) [inline tests/storage/lvm/test_lvm_sr.py:2]: This kind of cleanup should be in a separate commit.

File: tests/storage/lvmoiscsi/test_lvmoiscsi_sr.py [hunk 3]
```diff
@@ -1,5 +1,4 @@
 import pytest
-import subprocess
```
Comment (stormi) [inline tests/storage/lvmoiscsi/test_lvmoiscsi_sr.py:2]: idem

File: tests/storage/moosefs/conftest.py [hunk 4]
```diff
@@ -6,7 +6,8 @@ def host_with_moosefs(host):
         "MooseFS client should not be installed on the host before all tests"
     host.ssh(['sh', '-c', '"curl https://ppa.moosefs.com/RPM-GPG-KEY-MooseFS > /etc/pki/rpm-gpg/RPM-GPG-KEY-MooseFS"'])
     host.ssh(['sh', '-c', '"curl http://ppa.moosefs.com/MooseFS-3-el7.repo > /etc/yum.repos.d/MooseFS.repo"'])
-    host.yum_install(['fuse'], save_state=True)
+    host.yum_save_state()
+    host.yum_install(['fuse'])
```
Comment (stormi) [inline tests/storage/moosefs/conftest.py:10]: This should belong to the second commit.

File: tests/storage/storage.py [hunk 5]
```diff
@@ -1,14 +1,13 @@
...
         sr = host.sr_create(sr_type, label, {}, verify=True)
-    except subprocess.CalledProcessError as e:
+    except SSHCommandFailed as e:
         assert e.stdout == (
-            b'Error code: SR_BACKEND_FAILURE_90\nError parameters: , '
-            b'The request is missing the device parameter, \n'
-        ), 'Bad error, current: {}'.format(e.stdout)
+            'Error code: SR_BACKEND_FAILURE_90\nError parameters: , '
+            + 'The request is missing the device parameter, \n'
+        )
```
Comment (stormi) [inline tests/storage/storage.py:10]: This change (removal of useless assert error message) should be mentioned in the commit message.

File: tests/storage/nfs/test_nfs_sr.py [hunk 6]
```diff
@@ -1,7 +1,6 @@
 import pytest
 from lib.common import wait_for, vm_image
 import time
-import subprocess
```
Comment (stormi) [inline tests/storage/nfs/test_nfs_sr.py:4]: separate commit?

**Review by stormi — COMMENTED** [review]
I'm good with the SSHResult and SSHCommandFailed addition, as long as the output still contains all relevant information in case of test failure.

Good also about the save state changes, that make things more symmetrical and logical.

About the banner removal, I'm not convinced.

And my usual general comment, though I know it's still a WIP PR: commit messages that describe what they do and why :)

File: lib/common.py [hunk 7]
```diff
@@ -57,16 +56,28 @@ def xo_cli(action, args={}, check=True, simple_output=True):
         check=check
     )
     if simple_output:
-        return res.stdout.decode().strip()
+        return res.stdout.strip()
```
[the remark this answers is not in the document]
Comment (Wescoeur) [reply lib/common.py]: It's an error, I reverted this change. :)

File: lib/common.py [hunk 8]
```diff
@@ -131,8 +157,11 @@ def scp(hostname_or_ip, src, dest, check=True, suppress_fingerprint_warnings=Tru
         shell=True,
         stdout=subprocess.PIPE,
         stderr=subprocess.STDOUT,
-        check=check
+        check=False
     )
+    if check and res.returncode:
```
[the remark this answers is not in the document]
Comment (Wescoeur) [reply lib/common.py:166]: I don't think implicit conversion are a bad thing, I prefer a consise code. Generally a program returns 0 when it terminates correctly on Linux, IMH this code is readable. And But it's purely subjective.

File: lib/common.py [hunk 9]
```diff
@@ -131,8 +157,11 @@ def scp(hostname_or_ip, src, dest, check=True, suppress_fingerprint_warnings=Tru
...
+        raise SSHCommandFailed(res.returncode, res.stdout.decode())
```
[the remark this answers is not in the document]
Comment (Wescoeur) [reply lib/common.py]: Yeah, It was my point of view, do you think it's better to use another error class?

File: lib/common.py [hunk 10]
```diff
@@ -92,26 +103,41 @@ def ssh(hostname_or_ip, cmd, check=True, simple_output=True, suppress_fingerprin
...
-            check=check
+            check=False
         )
 
         # Even if check is False, we still raise in case of return code 255, which means a SSH error.
         if res.returncode == 255:
-            raise SSHException("SSH Error: %s" % res.stdout.decode())
+            raise SSHCommandFailed(255, "SSH Error: %s" % res.stdout.decode())
+        if banner_res.returncode == 255:
+            raise SSHCommandFailed(255, "SSH Error: %s" % banner_res.stdout.decode())
```
[the remark this answers is not in the document]
Comment (Wescoeur) [reply lib/common.py]: I'm not sure it's important to know the exact source of the error, when you get a 255 error, there is probably a connection problem or not valid route.

File: lib/common.py [hunk 11]
```diff
@@ -92,26 +103,41 @@ def ssh(hostname_or_ip, cmd, check=True, simple_output=True, suppress_fingerprin
...
         # Common case
+
+        # Fetch banner and remove it to avoid stdout/stderr pollution.
+        banner_res = subprocess.run(
+            "ssh root@%s %s '%s'" % (hostname_or_ip, ' '.join(options), '\n'),
+            shell=True,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            check=False
+        )
```
[the remark this answers is not in the document]
Comment (Wescoeur) [reply lib/common.py:122]: Do you have an or many examples regarding the potential problems with the double ssh call?

File: lib/common.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (stormi) [reply lib/common.py]: No. I'm not sure SSHCommandFailed is better than subprocess.CalledProcessError here, but I'm fine with both.

File: lib/common.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (stormi) [reply lib/common.py:122]: No, I have no specific examples, but I don't think it's good to do twice the amount of SSH calls just for an issue that only affected altered XCP-ng hosts, to my knowledge. If there was a better solution to filter out the headers, that would be fine, but here it seems too invasive a change to me.

File: lib/common.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (stormi) [reply lib/common.py:166]: I agree about things being concise when it's obvious, but I don't think your conversion of return code to bool where "True" means "there was an error" is that obvious. One has to remember that the return code is from the process and that 0 means success, and then to remember that the same 0 means False to python. Not everyone will get it without a bit of thinking.

[the remark this answers is not in the document]
Comment (stormi) [reply lib/common.py:166]: I kind of remember C/C++ projects where they created explicit macros to convert return codes to bool, e.g. SUCCESS(returncode) would evaluate to True if it's 0. There it becomes unambiguous.

File: lib/common.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (stormi) [reply lib/common.py]: Indeed, I missed the fact that it's only for error 255.

File: lib/common.py [hunk 12]
```diff
@@ -65,8 +64,23 @@ def xo_object_exists(uuid):
     lst = json.loads(xo_cli('--list-objects', {'uuid': uuid}))
     return len(lst) > 0
 
-class SSHException(Exception):
-    pass
+class SSHCommandFailed(Exception):
+    __slots__ = 'returncode', 'stdout'
```
Comment (stormi) [inline lib/common.py:68]: not needed anymore since you define the attributes in `__init__`? Else `cmd` is missing.

Comment (Wescoeur) [reply lib/common.py:68]: `__slots__` is useful to increase the performance when a property is accessed. I often use this attribute when the class is small. But yes, I must add `cmd` here. :)

Comment (stormi) [reply lib/common.py:68]: The docs I read suggest this performance improvement is mostly useful when there are going to be a lot of objects (thousands, millions), so now we're ready for some intensive testing!

Comment (Wescoeur) [reply lib/common.py]: Yes, but the another benefice is the documentation. I don't like the dynamic declaration of attributes in this language. :)

File: lib/common.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Wescoeur) [reply lib/common.py]: So I added an option to ignore or not the banner: `'--ignore-ssh-banner` in a separate commit.
(Unfortunately I couldn't find any other solution to ignore the banner without making two ssh calls.)

**Review by stormi — APPROVED** [review]
